Fixed error on Python3

dhanar10 opened the pull request:
I know that in nickovs/slimDNS#4 and nickovs/slimDNS#5 this proposal by @nevercast has been rejected because we want as small footprint as possible for MicroPython.

However, my use case is to use this as lightweight replacement for avahi-daemon in Raspbian on Raspberry Pi. 

This library can also be potentially useful for enhancing https://github.com/dapperfu/python_mDNServer so that it does not depend on avahi-daemon anymore.

I guess I will just leave this pull request here for anyone who is interested.

dhanar10 committed:
Fixed struct.error on Python 3

Fixes nickovs/slimDNS#4

dhanar10 committed:
Fixed ticks_diff, ticks_ms on Python 3

Fixes nickovs/slimDNS#5

nickovs commented:
Thanks for that. It seems that there is interest in supporting CPython so I'm happy to consider this. Let me run some tests on MicroPython to see how much difference it makes to the bytecode size in practice.

dhanar10 committed:
XXX Always multicast the reply

Faked handling of unicast replies seems to cause replies not reaching its destination in some cases

dhanar10 committed:
Added main function for Python3

Used for standalone execution

nickovs commented on slimDNS.py:
```diff
@@ -366,3 +376,12 @@ def test():
+    local_addr = sys.argv[1]
+    local_host = sys.argv[2]
+    server = SlimDNSServer(local_addr, local_host)
+    server.run_forever()
```
I'm happy to have a command-line wrapper for the library in a separate file but in the interests of size I'd like to keep this out of the main code.

nickovs commented on slimDNS.py:
```diff
@@ -291,7 +297,11 @@ def process_packet(self, buf, addr):
+        
+        # XXX Always multicast the reply. Faked handling of unicast replies
+        # seems to cause replies not reaching its destination in some cases.
+        self.sock.sendto(buf[:o], (_MDNS_ADDR, _MDNS_PORT))
```
Actually I'm sort of surprised that the old code worked at all, since it looks like I was comparing the source *IP address* to `_MDNS_PORT`. Can you try:
```
self.sock.sendto(buf[:o], (_MDNS_ADDR, _MDNS_PORT) if addr[1] == _MDNS_PORT else addr)
```
and see if it fixes the problem you were having?

dhanar10 replied:
Changing addr[0] to addr[1] as you suggested fixes the problem I am having.

nickovs commented on slimDNS.py:
```diff
@@ -23,10 +23,16 @@
     const = lambda x:x
-
-import time
+    ticks_ms = lambda : time.clock() * 1000.
+    ticks_diff = lambda a, b: b - a
+else:
+    ticks_ms = time.ticks_ms
+    ticks_diff = time.ticks_diff
+    
```
OK. This does not seem like undue overhead to support CPython.

dhanar10 committed:
Fixed handling of unicast replies

_MDNS_PORT was wrongly compared with source IP address

dhanarfmlx committed:
Separate command line wrapper from the library

dhanar10 committed:
Separate command line wrapper from the library

dhanar10 committed:
Merge branch 'master' of https://github.com/dhanar10/slimDNS

dhanar10 commented:
I have separated the command line wrapper and I have tested that the fix for unicast reply is working correctly. Do you have any other input on this pull request?

dhanar10 commented:
On second thought, does it really make sense from your point of view to accept this pull request if it contains the command line wrapper? If not, as long as the fix for python 3 & the unicast reply is accepted, I can just revert the command line wrapper and create my own project which depend on this project and contains the command line wrapper only.

dhanar10 committed:
Fix typo in command line wrapper

dhanar10 committed:
Fix yet another typo in command line wrapper

nickovs commented:
I think that the command line wrapper in a separate file is just fine here. When used in a MicroPython setting it will typically get copied onto the microcontroller manually anyway.
